lib/SILOptimizer/Utils/PartialApplyCombiner.cpp

```diff
@@ -229,78 +139,80 @@ bool PartialApplyCombiner::processSingleApply(FullApplySite paiAI) {
   // pre-incremented. When we combine the partial_apply and this apply into
   // a new apply we need to retain all of the closure non-address type
   // arguments.
+  auto destroyloc = RegularLocation::getAutoGeneratedLocation();
   auto paramInfo = pai->getSubstCalleeType()->getParameters();
   auto partialApplyArgs = pai->getArguments();
-  // Set of arguments that need to be destroyed after each invocation.
-  SmallVector<SILValue, 8> toBeDestroyedArgs;
   for (unsigned i : indices(partialApplyArgs)) {
-    auto arg = partialApplyArgs[i];
+    SILValue arg = partialApplyArgs[i];
+    if (argToTmpCopy.count(arg))
+      arg = argToTmpCopy.lookup(arg);
 
-    if (!arg->getType().isAddress()) {
+    if (paramInfo[paramInfo.size() - partialApplyArgs.size() + i]
+            .isConsumed()) {
       // Copy the argument as the callee may consume it.
-      arg = builder.emitCopyValueOperation(pai->getLoc(), arg);
-      // For non consumed parameters (e.g. guaranteed), we also need to
-      // insert destroys after each apply instruction that we create.
-      if (!paramInfo[paramInfo.size() - partialApplyArgs.size() + i]
-               .isConsumed())
-        toBeDestroyedArgs.push_back(arg);
+      if (arg->getType().isAddress()) {
+        auto *ASI = builder.createAllocStack(pai->getLoc(), arg->getType());
+        builder.createCopyAddr(pai->getLoc(), arg, ASI, IsTake_t::IsNotTake,
+                               IsInitialization_t::IsInitialization);
+        paiAI.insertAfterFullEvaluation([&](SILBasicBlock::iterator insertPt) {
+          SILBuilderWithScope builder(insertPt);
+          builder.createDeallocStack(destroyloc, ASI);
+        });
+        arg = ASI;
+      } else {
+        arg = builder.emitCopyValueOperation(pai->getLoc(), arg);
+      }
     }
+    // Add the argument of the partial_apply.
+    argList.push_back(arg);
   }
 
-  auto callee = pai->getCallee();
+  SILValue callee = pai->getCallee();
   SubstitutionMap subs = pai->getSubstitutionMap();
 
   // The partial_apply might be substituting in an open existential type.
   builder.addOpenedArchetypeOperands(pai);
 
-  FullApplySite nai;
-  if (auto *tai = dyn_cast<TryApplyInst>(paiAI))
-    nai = builder.createTryApply(paiAI.getLoc(), callee, subs, argList,
-                                 tai->getNormalBB(), tai->getErrorBB());
-  else
-    nai = builder.createApply(paiAI.getLoc(), callee, subs, argList,
-                              cast<ApplyInst>(paiAI)->isNonThrowing());
-
+  if (auto *tai = dyn_cast<TryApplyInst>(paiAI)) {
```

Comment:
Can you change this to be a full switch over the type of full apply site with llvm_unreachable in the cases which we do not support today. That way people know to update this code in the future.